File: CPP/ByRound/1955/1955_H_The_Most_Reckless_Defense.cpp
// Problem: CF 1955 H - The Most Reckless Defense
// https://codeforces.com/contest/1955/problem/H

/*
 * Problem: H. The Most Reckless Defense
 * 
 * This problem involves finding the maximum base health 'h' for which an enemy can
 * pass through a path in a tower defense game without dying, considering that towers
 * can be upgraded with ranges that increase the enemy's health by 3^r units.
 * 
 * Algorithm/Techniques:
 * - Use Kuhn-Munkres (Hungarian) algorithm for maximum weight bipartite matching
 * - For each possible range r (1 to 12), calculate how many path cells each tower attacks
 * - Assign weights based on (damage * count_of_cells) - 3^r to optimize the health gain
 * - The maximum weight matching gives the optimal assignment of ranges to towers
 * 
 * Time Complexity: O(R^3 * k), where R is max range (12) and k is number of towers
 * Space Complexity: O(R * k), for storing the bipartite graph and matching
 */

#include <assert.h>
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;
using ll = long long;
using db = double;
constexpr int mod = 998244353;
constexpr int N = 2e5 + 10;
struct KM {
  int n, m;
  vector<vector<ll>> a;
  vector<ll> lx, ly, mind;
  vector<int> px, py, pre;
  vector<bool> used;
  const ll inf = 1e18;
  KM(int _n, int _m) : n(_n), m(_m) {
    assert(n <= m);
    a = vector(n + 1, vector<ll>(m + 1));
    lx = vector<ll>(n + 1);
    ly = vector<ll>(m + 1);
    mind = vector<ll>(m + 1);
    px = vector<int>(n + 1, -1);
    py = vector<int>(m + 1, -1);
    pre = vector<int>(m + 1, -1);
    used = vector<bool>(m + 1);
  }
  void add_edge(int u, int v, ll w) { a[u][v] = max(a[u][v], w); }
  ll solve() {
    for (int i = 1; i <= n; i++) {
      fill(mind.begin(), mind.end(), inf);
      fill(used.begin(), used.end(), false);
      px[i] = 0;
      py[0] = i;
      int y = 0;
      while (true) {
        used[y] = true;
        int x = py[y];
        ll d = inf;
        int yy = -1;
        for (int j = 1; j <= m; j++) {
          if (used[j])
            continue;
          ll cur = lx[x] + ly[j] - a[x][j];
          if (mind[j] > cur) {
            mind[j] = cur;
            pre[j] = y;
          }
          if (d > mind[j])
            d = mind[j], yy = j;
        }
        for (int j = 0; j <= m; j++) {
          if (used[j])
            lx[py[j]] -= d, ly[j] += d;
          else
            mind[j] -= d;
        }
        y = yy;
        if (py[y] == -1)
          break;
      }
      for (; y; y = pre[y]) {
        py[y] = py[pre[y]];
        px[py[y]] = y;
      }
    }
    ll ans = 0;
    for (int i = 1; i <= n; i++) {
      if (px[i] != -1)
        ans += a[i][px[i]];
    }
    return ans;
  }
};
void solve() {
  int n, m, k;
  cin >> n >> m >> k;
  vector enemy(n + 1, vector<bool>(m + 1));
  for (int i = 1; i <= n; i++) {
    for (int j = 1; j <= m; j++) {
      char ch;
      cin >> ch;
      if (ch == '#')
        enemy[i][j] = true;
    }
  }
  vector<int> p(k + 1);
  vector<pair<int, int>> tower(k + 1);
  for (int i = 1; i <= k; i++) {
    cin >> tower[i].first >> tower[i].second >> p[i];
  }
  constexpr int R = 12;
  KM km(R, max(k, R));
  ll pow3 = 1;
  for (int r = 1; r <= R; r++) {
    pow3 *= 3;
    for (int i = 1; i <= k; i++) {
      auto [x, y] = tower[i];
      int cnt = 0;
      for (int dx = -r; dx <= r; dx++) {
        for (int dy = -r; dy <= r; dy++) {
          if (dx * dx + dy * dy <= r * r) {
            int x2 = x + dx;
            int yy = y + dy;
            if (1 <= x2 && x2 <= n && 1 <= yy && yy <= m) {
              if (enemy[x2][yy])
                cnt++;
            }
          }
        }
      }
      km.add_edge(r, i, p[i] * cnt - pow3);
    }
  }
  cout << km.solve() << "\n";
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int tt = 1;
  cin >> tt;
  while (tt--) {
    solve();
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1955/1955_H_The_Most_Reckless_Defense.py
# Problem: CF 1955 H - The Most Reckless Defense
# https://codeforces.com/contest/1955/problem/H

"""
Code Purpose:
This code solves the "The Most Reckless Defense" problem, which involves
optimizing tower ranges in a Tower Defense game to maximize the base health
of enemies that can pass through a path without dying. Each tower can have
a range from 1 to 10, and when set, it increases enemy health by 3^r, but each
r can only be used once. The goal is to determine the maximum base health h
such that the player still wins with this configuration.

Algorithms/Techniques:
- Dynamic Programming with Bitmasking
- Preprocessing tower coverage for different ranges
- Optimization to find best combination of tower ranges

Time Complexity: O(N * M * K * 10 + 2^10 * 10)
Space Complexity: O(N * M * K + 2^10)
"""

for _ in range(int(input())):
    N, M, K = map(int, input().split())
    Map = [list(input()) for _ in range(N)]

    A = []
    Dir = [(1, 0), (0, 1), (-1, 0), (0, -1)]
    for _ in range(K):
        a, b, c = map(int, input().split())
        B = [0] * 11
        a -= 1
        b -= 1

        # Precompute for each range how many cells in the path are covered
        for i in range(N):
            for j in range(M):
                if Map[i][j] == ".":
                    continue
                for k in range(11):
                    # Check if distance squared <= range squared
                    if (a - i) ** 2 + (b - j) ** 2 <= k**2:
                        B[k] += 1
                        break

        # Accumulate coverage to include all ranges up to current 
        for i in range(2, 11):
            B[i] += B[i - 1]
        # Multiply by tower power for each range
        for i in range(1, 11):
            B[i] *= c

        # Store only the first 10 ranges, starting from range 1
        A.append(B[1:])

    # DP with bitmasking to represent which tower ranges are chosen
    P = [0] * (2**10)

    # Iterate through all towers
    for data in A:
        Q = P.copy()

        # For each existing combination of ranges, try adding a new tower
        for i in range(2**10):
            for j in range(10):
                # Check if range j is not already used
                if i & (1 << j) == 0:
                    # Update the maximum damage for new combination
                    Q[i | (1 << j)] = max(Q[i | (1 << j)], P[i] + data[j])

        P = Q.copy()

    Ans = 0
    # Try all combinations of range assignments
    for i in range(2**10):
        n = i
        cnt = 0
        # Calculate the health increase from chosen ranges
        for j in range(10):
            cnt += (n % 2) * (3 ** (j + 1))
            n //= 2

        # Update answer with max damage minus health increase
        Ans = max(Ans, P[i] - cnt)

    print(Ans)


# https://github.com/VaHiX/CodeForces/